From 8fe1bf90f711c28e658e4fb66681b56f842ba3fd Mon Sep 17 00:00:00 2001
From: Robert Chiras <robert.chiras@nxp.com>
Date: Thu, 19 Apr 2018 11:22:08 +0300
Subject: [PATCH 3615/5242] MLK-18106: drm/imx: Fix nwl_dsi-imx bridge
 handling

commit  03c3482b5561725d2e4503255dc9143d6fcc88e2 from
https://source.codeaurora.org/external/imx/linux-imx.git

The nwl_dsi-imx which is the platform driver for the NortWest Logic DSI
host controller found on IMX platforms handles the DSI host as a bridge.
Since the drm_bridge_attach function now supports chained bridges, there
is no need for nwl_dsi_add_bridge and nwl_dsi_del_bridge functions.
Just pass the existent bridge to drm_bridge_attach.

This fixes a bug created during kernel 4.14 rebase process.

Signed-off-by: Robert Chiras <robert.chiras@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/nwl_dsi-imx.c |   12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/imx/nwl_dsi-imx.c b/drivers/gpu/drm/imx/nwl_dsi-imx.c
index 8a2e989..e97b174 100644
--- a/drivers/gpu/drm/imx/nwl_dsi-imx.c
+++ b/drivers/gpu/drm/imx/nwl_dsi-imx.c
@@ -702,13 +702,10 @@ static int imx_nwl_dsi_bridge_attach(struct drm_bridge *bridge)
 	}
 
 	/* Attach the next bridge in chain */
-	nwl_dsi_add_bridge(encoder, dsi->next_bridge);
-	ret = drm_bridge_attach(encoder, dsi->next_bridge, NULL);
-	if (ret) {
+	ret = drm_bridge_attach(encoder, dsi->next_bridge, bridge);
+	if (ret)
 		DRM_DEV_ERROR(dsi->dev, "Failed to attach bridge! (%d)\n",
 			      ret);
-		nwl_dsi_del_bridge(encoder, dsi->next_bridge);
-	}
 
 	return ret;
 }
@@ -719,7 +716,10 @@ static void imx_nwl_dsi_bridge_detach(struct drm_bridge *bridge)
 
 	DRM_DEV_DEBUG_DRIVER(dsi->dev, "id = %s\n",
 			     (dsi->instance)?"DSI1":"DSI0");
-	nwl_dsi_del_bridge(dsi->next_bridge->encoder, dsi->next_bridge);
+	/*
+	 * The next bridge in chain will be automatically detached, there is
+	 * no need for us to detach it.
+	 */
 }
 
 static const struct drm_bridge_funcs imx_nwl_dsi_bridge_funcs = {
-- 
1.7.9.5

